samples/drivers/i2s/output/src/main.c

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2023 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/i2s.h>
+#include <zephyr/sys/iterable_sections.h>
+
+#define SAMPLE_NO 64
+
+/* The data_l represent a sine wave */
+static int16_t data_l[SAMPLE_NO] = {
+	  3211,   6392,   9511,  12539,  15446,  18204,  20787,  23169,
+	 25329,  27244,  28897,  30272,  31356,  32137,  32609,  32767,
+	 32609,  32137,  31356,  30272,  28897,  27244,  25329,  23169,
+	 20787,  18204,  15446,  12539,   9511,   6392,   3211,      0,
+	 -3212,  -6393,  -9512, -12540, -15447, -18205, -20788, -23170,
+	-25330, -27245, -28898, -30273, -31357, -32138, -32610, -32767,
+	-32610, -32138, -31357, -30273, -28898, -27245, -25330, -23170,
+	-20788, -18205, -15447, -12540,  -9512,  -6393,  -3212,     -1,
+};
+
+/* The data_r represent a sine wave shifted by 90 deg to data_l sine wave */
+static int16_t data_r[SAMPLE_NO] = {
+	 32609,  32137,  31356,  30272,  28897,  27244,  25329,  23169,
+	 20787,  18204,  15446,  12539,   9511,   6392,   3211,      0,
+	 -3212,  -6393,  -9512, -12540, -15447, -18205, -20788, -23170,
+	-25330, -27245, -28898, -30273, -31357, -32138, -32610, -32767,
+	-32610, -32138, -31357, -30273, -28898, -27245, -25330, -23170,
+	-20788, -18205, -15447, -12540,  -9512,  -6393,  -3212,     -1,
+	  3211,   6392,   9511,  12539,  15446,  18204,  20787,  23169,
+	 25329,  27244,  28897,  30272,  31356,  32137,  32609,  32767,
+};
+
+
+static void fill_buf(int16_t *tx_block, int att)
+{
+	for (int i = 0; i < SAMPLE_NO; i++) {
+		tx_block[2 * i] = data_l[i] >> att;
+		tx_block[2 * i + 1] = data_r[i] >> att;
+	}
+}
+
+#define NUM_BLOCKS 20
+#define BLOCK_SIZE (2 * sizeof(data_l))
+
+#ifdef CONFIG_NOCACHE_MEMORY
+	#define MEM_SLAB_CACHE_ATTR __nocache
+#else
+	#define MEM_SLAB_CACHE_ATTR
+#endif /* CONFIG_NOCACHE_MEMORY */
+
+const struct device *dev_i2s = DEVICE_DT_GET(DT_ALIAS(i2s_tx));
+
+char MEM_SLAB_CACHE_ATTR __aligned(WB_UP(32))
+	_k_mem_slab_buf_tx_0_mem_slab[(NUM_BLOCKS) * WB_UP(BLOCK_SIZE)];
+
+STRUCT_SECTION_ITERABLE(k_mem_slab, tx_0_mem_slab) =
+	Z_MEM_SLAB_INITIALIZER(tx_0_mem_slab, _k_mem_slab_buf_tx_0_mem_slab,
+				WB_UP(BLOCK_SIZE), NUM_BLOCKS);
+
+int main(void)
+{
+	void *tx_block[NUM_BLOCKS];
+	struct i2s_config i2s_cfg;
+	int ret;
+	uint32_t tx_idx;
+
+	if (!device_is_ready(dev_i2s)) {
+		printf("I2S device not ready\n");
+		return -ENODEV;
+	}
+	/* Configure I2S stream */
+	i2s_cfg.word_size = 16U;
+	i2s_cfg.channels = 4U;
+	i2s_cfg.format = I2S_FMT_DATA_FORMAT_I2S;
```

Comment:
Is it anywhere documented the little-endian/big-endian convention within memory slab block? Here you just use the host endian, which is likely little-endian but should the code actually ensure it always writes to memory in little-endian? For things where the samples originate within the application (like here) it can be undefined, but when the samples do come from external sources then the format should really account that.

Reply:
I believe that data sent in the I2S format is specified to be MSB first, but I'm not sure there is any more beyond this: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/i2s.h#L47. @anangl are there additional configuration options I'm missing to set the endianness of I2S data?

Reply:
@anangl ping on the I2S format specification. @tmon-nordic, would you be ok with the PR moving forward as is? I'd like to try to get this sample merged if possible.

Reply:
> I believe that data sent in the I2S format is specified to be MSB first

Yes, but its the output from the I2S hardware and @tmon-nordic referred to data that is the input to that hardware. And that in my opinion should use the host endianness, as the sample does now. If the I2S hardware has a specific requirements in this regard (e.g. if some external I2S chip is used), its driver should take care of doing the required conversion.